dissect/esedb/btree.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from dissect.esedb.exceptions import KeyNotFoundError, NoNeighbourPageError
+
+if TYPE_CHECKING:
+    from dissect.esedb.esedb import EseDB
+    from dissect.esedb.page import Node, Page
+
+
+class BTree:
+    """A simple implementation for searching the ESE B+Trees.
+
+    This is a stateful interactive class that moves an internal cursor to a position within the BTree.
+
+    Args:
+        esedb: An instance of :class:`~dissect.esedb.esedb.EseDB`.
+        page: The page to open a BTree on.
+    """
+
+    def __init__(self, esedb: EseDB, root: int | Page):
+        self.esedb = esedb
+
+        if isinstance(root, int):
+            page_num = root
+            root = esedb.page(page_num)
+        else:
+            page_num = root.num
+
+        self.root = root
+
+        self._page = root
+        self._page_num = page_num
+        self._node_num = 0
+
+    def reset(self) -> None:
+        """Reset the internal state to the root of the BTree."""
+        self._page = self.root
+        self._page_num = self._page.num
+        self._node_num = 0
+
+    def node(self) -> Node:
+        """Return the node the BTree is currently on."""
+        return self._page.node(self._node_num)
+
+    def next(self) -> Node:
+        """Move the BTree to the next node and return it.
+
+        Can move the BTree to the next page as a side effect.
+        """
+        if self._node_num + 1 > self._page.node_count - 1:
+            self.next_page()
+        else:
+            self._node_num += 1
+
+        return self.node()
+
+    def next_page(self) -> None:
+        """Move the BTree to the next page in the tree.
```

Comment:
```suggestion
        """Move the BTree to the next :class:`~dissect.esedb.page.Page` in the tree.
```

Reply:
Knowledge of the page class is not relevant here.